scylla/src/statement/prepared_statement.rs

```diff
@@ -335,6 +346,83 @@ pub enum PartitionKeyError {
+        // Iterate on values using sorted pk_indexes (see deser_prepared_metadata),
+        // and use PartitionKeyIndex.sequence to insert the value in pk_values with the correct order.
+        let mut pk_values: SmallVec<[_; PartitionKey::SMALLVEC_ON_STACK_SIZE]> =
+            smallvec![None; prepared_metadata.pk_indexes.len()];
+        let mut values_iter = bound_values.iter();
+        // pk_indexes contains the indexes of the partition key value, so the current offset of the
+        // iterator must be kept, in order to compute the next position of the pk in the iterator.
+        // At each iteration values_iter.nth(0) will roughly correspond to values[values_iter_offset],
+        // so values[pk_index.index] will be retrieved with values_iter.nth(pk_index.index - values_iter_offset)
+        let mut values_iter_offset = 0;
+        for pk_index in prepared_metadata.pk_indexes.iter().copied() {
+            // Find value matching current pk_index
+            let next_val = values_iter
+                .nth((pk_index.index - values_iter_offset) as usize)
+                .ok_or_else(|| {
+                    PartitionKeyError::NoPkIndexValue(pk_index.index, bound_values.len())
+                })?;
+            // Add it in sequence order to pk_values
+            if let Some(v) = next_val {
+                let spec = &prepared_metadata.col_specs[pk_index.index as usize];
+                pk_values[pk_index.sequence as usize] = Some((v, spec));
+            }
+            values_iter_offset = pk_index.index + 1;
+        }
+        Ok(Self {
+            pk_values,
+            partitioner_name,
+        })
+    }
+
+    pub(crate) fn iter(&self) -> impl Iterator<Item = PartitionKeyValue<'ps>> + Clone + '_ {
+        self.pk_values.iter().flatten().copied()
+    }
+
+    pub(crate) fn calculate_token(&self) -> Result<Token, PartitionKeyError> {
+        let mut partitioner_hasher = self.partitioner_name.build_hasher();
+        let mut pk_val_iter = self.iter().map(|(val, _spec)| val);
+        if let Some(first_value) = pk_val_iter.next() {
+            if let Some(second_value) = pk_val_iter.next() {
+                // Composite partition key case
+                for value in std::iter::once(first_value)
+                    .chain(std::iter::once(second_value))
+                    .chain(pk_val_iter)
+                {
+                    let v_len_u16: u16 = value
+                        .len()
+                        .try_into()
+                        .map_err(|_| PartitionKeyError::ValueTooLong(value.len()))?;
+                    partitioner_hasher.write(&v_len_u16.to_be_bytes());
+                    partitioner_hasher.write(value);
+                    partitioner_hasher.write(&[0u8]);
+                }
+            } else {
+                // Single-value partition key case
+                partitioner_hasher.write(first_value);
+            }
+        }
+
+        Ok(partitioner_hasher.finish())
+    }
+}
```

Comment:
It looks like you duplicate the logic related to handling single-value and composite cases with `PreparedStatement::compute_partition_key`. Can we abstract them away? E.g. you could pass a `FnMut` closure that gets fed with the bytes.